### PR TITLE
add Sa-Solver

### PR DESCRIPTION
**Pull request by lawrence-cj**
# What does this PR do?
Add Sa-Solver into diffusers lib. 
Cc: @patil-suraj. Really need your help to test it.

### CODE REVIEW

**Comment by sayakpaul** [comment]
Looping in @yiyixuxu here.

File: src/diffusers/schedulers/scheduling_sasolver.py [hunk 1]
```diff
@@ -0,0 +1,858 @@
+# Copyright 2023 Stanford University Team and The HuggingFace Team. All rights reserved.
```
[the remark this answers is not in the document]
Comment (lawrence-cj) [reply src/diffusers/schedulers/scheduling_sasolver.py:1]: OK, I will

**Comment by yiyixuxu** [comment]
awesome :) 
Can we see an example of how to use it with SD1.5 and SDXL?

**Comment by lawrence-cj** [comment]
No problem, I will add a test for you.

File: src/diffusers/schedulers/scheduling_sasolver.py [hunk 2]
```diff
@@ -0,0 +1,858 @@
...
+
+        """
+        if self.num_inference_steps is None:
+            raise ValueError(
+                "Number of inference steps is 'None', you need to run 'set_timesteps' after creating the scheduler"
+            )
+
+        if isinstance(timestep, torch.Tensor):
+            timestep = timestep.to(self.timesteps.device)
+        step_index = (self.timesteps == timestep).nonzero()
```
[the remark this answers is not in the document]
Comment (lawrence-cj) [reply src/diffusers/schedulers/scheduling_sasolver.py:748]: We implemented sa-solver according to DPMSolverMultistep in diffusers==0.21.2. Could you please help us to change it to the newest implementation?

[the remark this answers is not in the document]
Comment (yiyixuxu) [reply src/diffusers/schedulers/scheduling_sasolver.py:748]: sure :)

**Comment by sayakpaul** [comment]
There are some tests failing here. Could we please fix them?

**Comment by sayakpaul** [comment]
I will let @yiyixuxu review this once the test failures are resolved.

**Comment by patrickvonplaten** [comment]
Gentle ping here @yiyixuxu

**Comment by lawrence-cj** [comment]
fixed some error here. Cc: @sayakpaul, @yiyixuxu

**Comment by yiyixuxu** [comment]
seems the added fast test for sa-solver is failing?  @lawrence-cj

**Comment by lawrence-cj** [comment]
I was wandering [how to get the specific value](https://github.com/huggingface/diffusers/pull/5975#discussion_r1413759563) that is causing the error.

**Comment by lawrence-cj** [comment]
Test files passed. Gentle ping here. @sayakpaul @yiyixuxu

**Comment by lawrence-cj** [comment]
Gentle ping here. Cc: @yiyixuxu

File: src/diffusers/schedulers/scheduling_sasolver.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (patrickvonplaten) [reply src/diffusers/schedulers/scheduling_sasolver.py:766]: As stated by @yiyixuxu, we need to change the logic to use a `self.step_index` counter or else this scheduler won't function correctly when using karras sigmas or when doing img2img translation. Could you take a look at https://github.com/huggingface/diffusers/blob/f72b28c75b2b4b720a5d8de78556694cf4b893fd/src/diffusers/schedulers/scheduling_dpmsolver_multistep.py#L826 and refactor the code accordingly @lawrence-cj ?

**Comment by patrickvonplaten** [comment]
@lawrence-cj, I added a commit that should fix the 'make style' test. Could you check: https://github.com/huggingface/diffusers/pull/5975#discussion_r1442741053 and try to apply it? :-)

File: src/diffusers/schedulers/scheduling_sasolver.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (lawrence-cj) [reply src/diffusers/schedulers/scheduling_sasolver.py:848]: Cc: @scxue

**Review by yiyixuxu — COMMENTED** [review]
Thanks!

we are accepting use_karras_sigmas here so I think we need to update all our calculations to based on sigmas instead of timesteps in order to avoid the duplicated timestep issue

see relevant PRs:
https://github.com/huggingface/diffusers/pull/4986
https://github.com/huggingface/diffusers/pull/5158
https://github.com/huggingface/diffusers/pull/6006

happy to make it over now and help get this merged if it's ok with you? @lawrence-cj

**Comment by lawrence-cj** [comment]
We add `use_karras_sigmas` in the sa-solver in the newest commit. Could you please help to check the correctness. @yiyixuxu

**Comment by lawrence-cj** [comment]
Genle ping here. @yiyixuxu

**Comment by patrickvonplaten** [comment]
Happy to merge whenever @yiyixuxu is happy here!

File: src/diffusers/schedulers/scheduling_sasolver.py
Comment (lawrence-cj) [inline src/diffusers/schedulers/scheduling_sasolver.py:1]: I'm wondering how you make this style automatically. : )

**Comment by lawrence-cj** [comment]
I'm wondering how you make this style automatically. : ) Any tools for it?

**Comment by yiyixuxu** [comment]
@lawrence-cj if you installed our dev environment, you could run `make style` command :)
see our guide here https://huggingface.co/docs/diffusers/conceptual/contribution#how-to-open-a-pr

**Comment by lawrence-cj** [comment]
Oh, thank you so much for the information. @yiyixuxu

**Review by yiyixuxu — APPROVED** [review]
thanks!

**Comment by yiyixuxu** [comment]
@lawrence-cj 
thanks for adding this scheduler! I'm going to merge this in now.
we recently merged in this PR https://github.com/huggingface/diffusers/pull/6477 that improves the generation quality for DPM schedulers + SDXL - would same fix works for Sa-Solver? we can open an PR later to add same fix if this is the case!

**Pull request merged**